### ISSUE TITLE
Link error: Undefined reference to function caused by RepoPruningPass and followed GlobalDCELegacyPass.

### ISSUE THREAD

**Issue by MaggieYingYi** [SNSystems/llvm-project-prepo#76]
Consider two files: foo.c and main.c, listed below.

$ cat foo.c
~~~C
static int f (void) {
    static int e = 0;
    return e++;
}
int g (void) {
    return f () + 1;
}
~~~
$ cat main.c
~~~C
extern int g (void);
int main () {
    return g ();
}
~~~
Firstly, compile and link using the following commands:
~~~
clang -c -o foo.o -O0 -target x86_64-pc-linux-gnu-repo foo.c
repo2obj -o foo.nato foo.o
clang -c -o main.o -O0 -target x86_64-pc-linux-gnu-repo main.c
repo2obj -o main.nato main.o
clang   -o a.out foo.nato main.nato
~~~
build is fine. If we modified the initial value of variable `e` from `0` to `1` and recompile and link again, we got the following link error:
~~~
foo.nato: In function `g':
(.text+0x5): undefined reference to `f'
clang-8: error: linker command failed with exit code 1 (use -v to see invocation)
~~~

**Comment by MaggieYingYi** [comment]
**Analysis:**

When we compile the source code with the flag `-O0`, there is `noinline` attribute for all functions. Therefore, when calculating a function hash, its hash will not accumulate its dependents’ hash. The final function hash is the hash of the function itself.

When we modify the variable `e` value, the hash value of `f` is changed. However, the hash value of its caller `g` keeps the same. During the second time compile, the function `g` is pruned and `f` is not pruned. Since the function `g` is pruned and has an empty `UseList`, it will be deleted. After repo pruning, the function `g` has been changed from `definition` to `declaration`, which will result in changing the `UseList` of function `f` from 1 to 0.

The Dead Global Elimination pass is after the repo pruning pass. Since the function `f` is local function and has an empty `UseList`, it is removed as well. Therefore, there is only one compilation member `g` for the file foo.c during the second build. There are two missing compilation members: function `f` and variable `f.e`. This causes the Link error.

**Comment by MaggieYingYi** [comment]
**Issue Description**

If the caller and callee satisfies the following conditions, the undefined reference to the callee will be given.

1. Caller is pruned and deleted during the RepoPruning pass.
2. Callee isn’t pruned, has noinline attribute and may be discarded if it is not used.

The reason is that the callee is removed during the Dead Global Elimination pass, which is after the repo pruning pass.  The current hash calculation algorithm doesn’t accumulate the callee's hash value to the caller if the callee has noinline attribute.

**Proposed Solution**

To solve the issue, the caller always accumulates the callee’s hash value if callee may be discarded if it is not used.

**Comment by MaggieYingYi** [comment]
The fix is committed into the master branch (c660495e).
